packages/contracts/contracts/voting/erc20/Erc20VotingManager.sol

```diff
@@ -0,0 +1,215 @@
+        view
+        virtual
+        override
+        returns (RequestedPermission[] memory permissions, string[] memory helperNames)
+    {
+        (, TokenSetting memory _tokenSetting) = abi.decode(data, (VoteSetting, TokenSetting));
+
+        permissions = new RequestedPermission[](_tokenSetting.addr == address(0) ? 5 : 4);
+        helperNames = new string[](_tokenSetting.addr == address(0) ? 1 : 0);
+
+        address NO_ORACLE = address(0);
+
+        // Grant the `EXECUTE_PERMISSION_ID` permission on the installing DAO to the plugin.
+        permissions[0] = buildPermission(
+            BulkPermissionsLib.Operation.Grant,
+            DAO_PLACEHOLDER,
+            PLUGIN_PLACEHOLDER,
+            NO_ORACLE,
+            keccak256("EXECUTE_PERMISSION")
+        );
+
+        // Grant the `SET_CONFIGURATION_PERMISSION_ID` permission on the plugin to the installing DAO.
+        permissions[1] = buildPermission(
+            BulkPermissionsLib.Operation.Grant,
+            PLUGIN_PLACEHOLDER,
+            DAO_PLACEHOLDER,
+            NO_ORACLE,
+            erc20VotingBase.SET_CONFIGURATION_PERMISSION_ID()
+        );
+
+        // Grant the `UPGRADE_PERMISSION_ID` permission on the plugin to the installing DAO.
+        permissions[2] = buildPermission(
+            BulkPermissionsLib.Operation.Grant,
+            PLUGIN_PLACEHOLDER,
+            DAO_PLACEHOLDER,
+            NO_ORACLE,
+            erc20VotingBase.UPGRADE_PERMISSION_ID()
+        );
+
+        // Grant the `SET_TRUSTED_FORWARDER_PERMISSION_ID` permission on the plugin to the installing DAO.
+        permissions[3] = buildPermission(
+            BulkPermissionsLib.Operation.Grant,
+            PLUGIN_PLACEHOLDER,
+            DAO_PLACEHOLDER,
+            NO_ORACLE,
+            erc20VotingBase.SET_TRUSTED_FORWARDER_PERMISSION_ID()
+        );
+
+        uint256 GOVERNANCE_ERC20_HELPER_IDX = 0;
+
+        if (_tokenSetting.addr == address(0)) {
+            // Grant the `MINT_PERMISSION_ID` permission on the`GovernanceErc20` helper contract with index `GOVERNANCE_ERC20_HELPER_IDX` to the installing DAO.
+            permissions[4] = buildPermission(
+                BulkPermissionsLib.Operation.Grant,
+                GOVERNANCE_ERC20_HELPER_IDX,
+                DAO_PLACEHOLDER,
+                NO_ORACLE,
+                GovernanceERC20(governanceERC20Base).MINT_PERMISSION_ID()
+            );
+
+            helperNames[GOVERNANCE_ERC20_HELPER_IDX] = "GovernanceERC20";
+        }
+    }
+
+    /// @inheritdoc PluginManager
+    function getImplementationAddress() public view virtual override returns (address) {
+        return address(erc20VotingBase);
+    }
+
+    /// @inheritdoc PluginManager
+    function deployABI() external view virtual override returns (string memory) {
+        return
+            "((tuple(uint64,uint64,uint64) voteSetting),(tuple(address addr,string name,string symbol, (tuple(address[],uint256[]) mintSetting)) tokenSetting))";
```

Comment:
Couldn't we make it flat and with names?
In terms of UI, drawing a recursive ABI-based form with unnamed tuple fields like this is going to be a no-go.

Reply:
thanks, will be fully tested in the testing task